cli/src/test/java/com/devonfw/tools/ide/commandlet/UpdateCommandletTest.java

```diff
@@ -56,4 +56,30 @@ private void deleteTemplatesFolder(IdeContext context) throws IOException {
     Files.delete(templates.getParent());
     Files.delete(templates.getParent().getParent());
   }
+
+  /**
+   * Tests if a sub step (installation of software) of update failed, the overall update process will not fail too.
+   * <p>
+   * See: <a href="https://github.com/devonfw/IDEasy/issues/628">#628</a> for reference.
+   */
+  @Test
+  public void testRunUpdateSoftwareDoesNotFailOnFailedSoftwareInstallations() {
+
+    // arrange
+    IdeTestContext context = newContext(PROJECT_UPDATE);
+    Path javaRepository = context.getToolRepositoryPath().resolve("default").resolve("java");
+    context.getFileAccess().delete(javaRepository);
+    Path javaDownload = context.getIdeRoot().resolve("repository").resolve("java");
+    context.getFileAccess().delete(javaDownload);
+    UpdateCommandlet uc = context.getCommandletManager().getCommandlet(UpdateCommandlet.class);
+
+    // act
+    uc.run();
+
+    // assert
+    assertThat(context).logAtError().hasMessage("Installation of java failed!");
```

Comment:
We could even assert on the step structure here via `this.context.getCurrentStep()`.
In `StepImpl` we could expose `getChildren()` for testing.
Just a thought since here we do not assert the status of aggregated steps nor the overall success or failure.
Also by thinking this further and just by reading the code I am now confident that there is an error and this will not work:
https://github.com/devonfw/IDEasy/blob/2e601cd6b0efcb7d6b77e88155543fa2e7cbb569/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java#L766-L779

First of all the overall status should not be success but error and we should get exit code != 0 in such case.
My assumption is even that we will fail with an assertion error if we succeed the root step if it has child steps that are not successful but I have not verified this. At least my expectation is that a step should only be successful if all its child steps are successful.
This is all not your fault and your PR and changes are great. It is just that this review reveals new related problems that I now get aware of. IMHO we should also fix that together. But if you want you can also create a new bug issue and merge this first...

Reply:
I've create a new issue now: #639

Reply:
Thanks